staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -592,17 +592,22 @@ func (c *Cacher) Watch(ctx context.Context, key string, opts storage.ListOptions
 		identifier,
 	)
 
+	// note that c.waitUntilWatchCacheFreshAndForceAllEvents must be called without
+	// the c.watchCache.RLock held otherwise we are at risk of a deadlock
+	// mainly because c.watchCache.processEvent method won't be able to make progress
```

Comment:
Please also add a comment that it is safe to release the lock after that as we don't require any atomicity between this call and further call that actually grabs the events.